### DISCUSSION TITLE
Should the Jamulus server use multiple UDP ports to combat jitter?

### DISCUSSION THREAD

**Discussion by hselasky** [jamulussoftware/jamulus#2912] [Ideas]
Hi,

I'm the author of HPSJAM, See https://github.com/hselasky/hpsjam . I recently made an interesting discovery, that the biggest enemy for low latency audio is Jitter and not packet loss.

The internet is connected not only using one physical connection, but many. The IPv4- and IPv6- address and UDP port numbers are used to select on which physical fiber your application traffic is put. Also inside routers, traffic is split up into multiple so-called per-CPU queues, using the exact same principles. When using fixed address information it is called a stream and packets are processed in sequence. When using variable address information it is called a multi-stream.

When many applications output data at the same time, the streams that end up on the same packet queues and physical fiber links may overflow at random points in time causing random Jitter. But not for all queues at the same time.

By using multi-UDP streams (15-ports) allowing up to one of three packets to get delayed or lost, low latency applications end up with much higher overall quality and stability I've found. In HPSJAM there is now a GUI button to select single- or multi- port and the difference is quite noticable.

I was wondering if Jamulus could do something similar? I've used 15 ports, because the sequence number spins around 255, because that makes sense for the 1/3 packet loss too.

Packet 1 goes to port+0, packet 2 goes to port+1 and so on, until it repeats.

Any comments?

--HPS

**Comment by ghost** [comment]
This is interesting research. I think it is good to keep this optional because there may be Pros and Cons depending on the situation. There may be other similar methods that work also. I 'just thought now' that the server could assign different unchanging ports for different clients for a similar effect. Note that Jamulus needs to be a good internet citizen, and not look menacing either, to avoid negative attention.

**Reply by hselasky** [reply]
Using 15 UDP ports per server is not any issue. There are plenty of UDP ports to take from!

**Comment by pljones** [comment]
That requires at least a 15 packet buffer length, to allow for out of sequence arrivals, to avoid loss, though, doesn't it?  If there's, say, three packets of random jitter on each queue additional buffering for that fifteen lots of jitter will get added on top of the 15 packets inherently required for sequencing.  So that adds up to more latency, surely?

**Comment by hselasky** [comment]
@pljones :

I just use 3x1ms HPSJAM packets for out of sequence arrivals. There are some magic constants in there.

It has something to do with statistics and randomness. It's not directly logical.

HPSJAM allows 1/3 of packets to dissappear, unlike Jamulus. Which means HPSJAM takes the first 10 packets out of 15, before it starts producing sound. If the 10 best links out of 15 have similar latency, it adds no more than 3 ms of delay on RX.

For example a 3 ms HPSJAM frame looks like this:
PKT 1 - data
PKT 2 - data
PKT 3 - redundancy (xored data 1 and 2)

Supposedly the three packets go on different links and we pick the 2 first packets (any of the 3) which allows for full audio reconstruction.

This repeats 5 times for the 15 ports.

--HPS

**Reply by pljones** [reply]
So you'd want a _minimum_ of three ports for three separate routes and 3x1ms for out of sequence buffering for that set of 3x1ms ports.  Each set of three ports would need its own 3x1 buffer..?  But then you still need to account for differences in timing between the sets of three ports.  So each increment in a set of 3x1ms streams would take up another 2ms buffer, in case you had to handle reordering across sets of streams, surely?  So for 5x3x1ms you get 5x3ms buffering plus 5x2ms buffer - 25ms buffer minimum, right?

Of course, because the 3x1ms buffers you don't care what order they fill, that's fixed.  But that set of three ports can start receiving the next 3x1ms of buffers before you've had the other four lots in.  And it could start getting the next 3x1ms of buffers after _that_, too...

At best, you get them in close to the right order, of course. :)  1+2+(1⊕2), 3+4+(3⊕4), ... and you only need the extra 3x1ms for jitter / out of order.  But 3 and 4 could get delayed until after 6 and 7, for instance...  So I'd still think you need to filter each resolved pair into a jitter buffer.  It may improve resiliency but I don't think it reduces jitter or latency.

**Reply by ann0see** [reply]
> It may improve resiliency but I don't think it reduces jitter or latency.

The question is if that's the goal. Sometimes it might be favorable to have better sound quality with higher latency, I'd say. The question is if the gain in audio quality is better with larger buffers or the multi port approach (and which leads to the higher latency)?

**Reply by pljones** [reply]
So would a single 150 channel server still only listen for packets on the one port or would it need 2550 ports?  Responses to each client would need 15 ports, so 2250 there.  And if I were running 15 servers on a single static IP, that would need 33750 ports (rather than 2550)?

**Reply by hselasky** [reply]
> So you'd want a _minimum_ of three ports for three separate routes and 3x1ms for out of sequence buffering for that set of 3x1ms ports. Each set of three ports would need its own 3x1 buffer..? But then you still need to account for differences in timing between the sets of three ports. So each increment in a set of 3x1ms streams would take up another 2ms buffer, in case you had to handle reordering across sets of streams, surely? So for 5x3x1ms you get 5x3ms buffering plus 5x2ms buffer - 25ms buffer minimum, right?

No, it is not exactly done like that.

There are 15 packets for 15 ports. One packet is sent every 1ms. But I don't wait for 15 ms of audio before I send those 15 packets.

In 3 ms there are 3 * 48 = 144 samples. This is split into two data packets of 72 samples.

On the TX there is a small audio buffer of a few milliseconds to even out that we don't pull audio on the redundancy packet.
Then packets are sent as we go, with very little time delay.

On the RX side there is some clever logic. The maximum time window is 255 ms.

Then I record, where do I see the first packet and where do I see the last packet. This in turn produces a time interval representing the difference in delay between the various routes. If the spread is big, it means more buffering. If the spread is low, it means less buffering.

Then I do a rough estimate and compute if I received 2 out of 3 packets in the given time interval, even if not all 2 out of 3 consequtive packets are received, I will start to pull audio from a point, X. Frames arriving too late, before point X are silently dropped. After I've pulled a frame point X is incremented. The window I look at is 255 ms divided by 2 which is 127 ms approx.

> 
> Of course, because the 3x1ms buffers you don't care what order they fill, that's fixed. But that set of three ports can start receiving the next 3x1ms of buffers before you've had the other four lots in. And it could start getting the next 3x1ms of buffers after _that_, too...
> 
> At best, you get them in close to the right order, of course. :) 1+2+(1⊕2), 3+4+(3⊕4), ... and you only need the extra 3x1ms for jitter / out of order. But 3 and 4 could get delayed until after 6 and 7, for instance... So I'd still think you need to filter each resolved pair into a jitter buffer. It may improve resiliency but I don't think it reduces jitter or latency.

You need to account that the 1+2+(1⊕2) and 3+4+(3⊕4) are all sent at different times, which must be subtracted.

You can try installing HPSJAM and setup a server / client pair. It's very identical to how Jamulus works. In the HPSJAM client there is a drop down list which allows you to select single port or multi port. Then hear the difference for yourself.

I will admit I use Jamulus too, and want it to be better.

**Reply by pljones** [reply]
> The maximum time window is 255 ms.

That's huge.  You can jam with people in Brazil or South Korea from London with that kind of latency.  (I've not tried it, I so I won't comment on packet loss.)

> The window I look at is 255 ms divided by 2 which is 127 ms approx.

Basically a sliding window over the 255ms buffer?

> You need to account that the 1+2+(1⊕2) and 3+4+(3⊕4) are all sent at different times, which must be subtracted.

I've understood that you can send 1+2+(1⊕2) as soon as it's happened, not wait for anything else.  I'm only really talking about what the receiver needs to do to cope with jitter (and the _much_ more rare loss of two complementary packets).

**Reply by ann0see** [reply]
> You can try installing HPSJAM and setup a server / client pair. It's very identical to how Jamulus works. In the HPSJAM client there is a drop down list which allows you to select single port or multi port. Then hear the difference for yourself.

Are there any official servers available?

> I will admit I use Jamulus too, and want it to be better.

;-). Certainly that's great to hear. If we introduce protocol changes, they should be mostly end user invisible (and as backward compatible as possible).

**Reply by hselasky** [reply]
> So would a single 150 channel server still only listen for packets on the one port or would it need 2550 ports? Responses to each client would need 15 ports, so 2250 there. And if I were running 15 servers on a single static IP, that would need 33750 ports (rather than 2550)?

Using HPSJAM _only_ the server side use multiple ports. The client side only use a single port. This is due to the firewall / NAT issue, then if the client use multiple UDP ports instead, then the server would need to track that, which is more overhead.

All HPSJAM clients send data to the same 15 server ports. When you connect you only give the first UDP port, and it is assumed that the other 15 are there. It is also assumed that the first connect packet goes on UDP port N+0.

This is also good for load balancing, that now you can receive UDP data on multiple threads, and not only at the application level, but also at the kernel level, you utilize more hardware queues. This is good for lower latency. I know that different OS'es behave differently there, but the principle is the same.

**Reply by pljones** [reply]
> Using HPSJAM only the server side use multiple ports. The client side only use a single port.

Ah right.  So the 15 ports on the server are shared by all clients and it's still just one response port per client.  OK.  (So only more ports than Jamulus server on the incoming side.)

**Reply by hselasky** [reply]
> 
> Are there any official servers available?
> 

Not really official server yet, but you can test these if you don't have server access:

5.onj.me:34343   (Located at server location in the UK)
home.selasky.org:700 (Located privately in Norway, This one also has a synth and reverb connected on MIDI channel 1, 2 and 3. Avoid program change events, because it's not really GM compatible :-)

--HPS

**Reply by hselasky** [reply]
> > Using HPSJAM only the server side use multiple ports. The client side only use a single port.
> 
> Ah right. So the 15 ports on the server are shared by all clients and it's still just one response port per client. OK. (So only more ports than Jamulus server on the incoming side.)

Yes, correct!

**Reply by hselasky** [reply]
> > > Using HPSJAM only the server side use multiple ports. The client side only use a single port.
> > 
> > 
> > Ah right. So the 15 ports on the server are shared by all clients and it's still just one response port per client. OK. (So only more ports than Jamulus server on the incoming side.)
> 
> Yes, correct!

To ensure no packets are lost in the firewall, I always respond from the UDP N+0 port on the server side the whole first second, while the client is banging N+0..N+15, so that I'm sure the firewall / NAT is ready to receive packets on those other ports aswell. Not a problem for IPv6 though.

**Comment by hselasky** [comment]
During the last few days I've made a slew of changes to HpsJAM w.r.t. multi-path UDP routing. If Jamulus could do something similar audioquality would improve instantly w/o having to sacrifice any bandwidth. I think you can use my techniques even without redundancy packets. Please give HpsJAM a spin and you'll see the difference. Sorry I don't have time to make patches for Jamulus.

**Reply by mcfnord** [reply]
"Audio quality would improve without sacrificing any bandwidth."

Is reduced jitter the result? 

Would this feature require server operators to open 15 more ports? Would opening these be as seamless as it is for just one port today for registered servers?

It sounds like clients would transmit on one or more of 16 ports, but it also sounds like that's not a big deal (no additional configuration).
